sql/core/src/main/scala/org/apache/spark/sql/execution/ExternalAppendOnlyUnsafeRowArray.scala

```diff
@@ -106,6 +107,8 @@ private[sql] class ExternalAppendOnlyUnsafeRowArray(
   def add(unsafeRow: UnsafeRow): Unit = {
     if (numRows < numRowsInMemoryBufferThreshold) {
       inMemoryBuffer += unsafeRow.copy()
+      numRows += 1
```

Comment:
I think that we need to increment `numRows` and `modificationsCount` in both branches, not just this one: if we only update it in this branch then `length()` will return the wrong result.

If you want to know how many rows are in the in-memory buffer then I'd define a separate `numRowBufferedInMemory` variable; that will also help to make the code in `MergerIterator` easier to understand.

Reply:
Yes, make sense. I will push new change in the coming days. Thank you